flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemCommitter.java

```diff
@@ -0,0 +1,161 @@
+ *   updates metadata, it will be faster, so the probability of inconsistency is relatively small.
+ *
+ * <p>See:
+ * {@link TempFileManager}.
+ * {@link FileSystemLoader}.
+ */
+@Internal
+public class FileSystemCommitter implements Serializable {
+
+	private static final long serialVersionUID = 1L;
+
+	private final FileSystemFactory factory;
+	private final MetaStoreFactory metaStoreFactory;
+	private final boolean overwrite;
+	private final Path tmpPath;
+	private final LinkedHashMap<String, String> staticPartitions;
+	private final int partitionColumnSize;
+
+	public FileSystemCommitter(
+			FileSystemFactory factory,
+			MetaStoreFactory metaStoreFactory,
+			boolean overwrite,
+			Path tmpPath,
+			LinkedHashMap<String, String> staticPartitions,
+			int partitionColumnSize) {
+		this.factory = factory;
+		this.metaStoreFactory = metaStoreFactory;
+		this.overwrite = overwrite;
+		this.tmpPath = tmpPath;
+		this.staticPartitions = staticPartitions;
+		this.partitionColumnSize = partitionColumnSize;
+	}
+
+	/**
+	 * For committing job's output after successful batch job completion or one checkpoint finish
+	 * for streaming job. Should move all files to final output paths.
+	 *
+	 * <p>NOTE: According to checkpoint notify mechanism of Flink, checkpoint may fail and be
+	 * abandoned, so this method should commit all checkpoint ids that less than current
+	 * checkpoint id (Includes failure checkpoints).
+	 */
+	public void commitUpToCheckpoint(long toCpId) throws Exception {
+		FileSystem fs = factory.create(tmpPath.toUri());
+
+		try (FileSystemLoader loader = new FileSystemLoader(overwrite, fs, metaStoreFactory)) {
+			for (long cp : headCheckpoints(fs, tmpPath, toCpId)) {
+				commitSingleCheckpoint(fs, loader, cp);
+			}
+		}
+	}
+
+	private void commitSingleCheckpoint(
+			FileSystem fs, FileSystemLoader loader, long checkpointId) throws Exception {
+		try {
+			List<Path> taskPaths = listTaskTemporaryPaths(fs, tmpPath, checkpointId);
+			if (partitionColumnSize > 0) {
+				for (Map.Entry<LinkedHashMap<String, String>, List<Path>> entry :
+						collectPartSpecToPaths(fs, taskPaths).entrySet()) {
+					loader.loadPartition(entry.getKey(), entry.getValue());
+				}
+			} else {
+				loader.loadNonPartition(taskPaths);
+			}
+		} finally {
+			deleteCheckpoint(fs, tmpPath, checkpointId);
+		}
+	}
+
+	/**
+	 * Create a new temporary file manager from task and checkpoint id.
+	 * And clean the temporary directory for task.
+	 */
+	public TempFileManager createFileManagerAndCleanDir(
+			int taskNumber, long checkpointId) throws Exception {
+		TempFileManager pathGenerator = new TempFileManager(
+				tmpPath, taskNumber, checkpointId);
+		Path path = pathGenerator.getTaskTemporaryPath();
+		factory.create(path.toUri()).delete(path, true);
+		return pathGenerator;
+	}
+
+	/**
+	 * Collect all partitioned paths, aggregate according to partition spec.
+	 */
+	private Map<LinkedHashMap<String, String>, List<Path>> collectPartSpecToPaths(
```

Comment:
This should move to `TempFileManager`, only it knows how files and directories are organized.